File: _CodeTopics/LeetCode_contest/biweekly/biweekly2021/48/48_2.py
class AuthenticationManager(object):

    def __init__(self, timeToLive):
        """
        :type timeToLive: int
        """
        self.mp = {}
        self.live = timeToLive

    def generate(self, tokenId, currentTime):
        """
        :type tokenId: str
        :type currentTime: int
        :rtype: None
        """
        self.mp[tokenId] = currentTime + self.live

    def renew(self, tokenId, currentTime):
        """
        :type tokenId: str
        :type currentTime: int
        :rtype: None
        """
        if tokenId in self.mp and currentTime < self.mp[tokenId]:
            self.mp[tokenId] = currentTime + self.live

    def countUnexpiredTokens(self, currentTime):
        """
        :type currentTime: int
        :rtype: int
        """
        res = 0
        for v in self.mp.values():
            if currentTime < v:
                res += 1
        return res


# Your AuthenticationManager object will be instantiated and called as such:
# obj = AuthenticationManager(timeToLive)
# obj.generate(tokenId,currentTime)
# obj.renew(tokenId,currentTime)
# param_3 = obj.countUnexpiredTokens(currentTime)

"""
https://leetcode-cn.com/submissions/detail/157686015/

90 / 90 个通过测试用例
状态：通过
执行用时: 544 ms
内存消耗: 14.3 MB
"""

File: _CodeTopics/LeetCode/1601-1800/001797--biweekly48_2/tran/48_2.cpp
class AuthenticationManager {
    
    // 第 48 场双周赛第二题
    // https://leetcode.cn/submissions/detail/157686015/

private:
    unordered_map<string, int> _mp;
    int _live;

public:
    AuthenticationManager(int timeToLive) {
        _mp = unordered_map<string, int>{};
        _live = timeToLive;
    }
    
    void generate(string tokenId, int currentTime) {
        _mp[tokenId] = currentTime + _live;
    }
    
    void renew(string tokenId, int currentTime) {
        if (_mp.find(tokenId) != _mp.end() && currentTime < _mp[tokenId]) {
            _mp[tokenId] = currentTime + _live;
        }
    }
    
    int countUnexpiredTokens(int currentTime) {
        int res = 0;
        for (auto [k, v] : _mp) {
            if (currentTime < v) {
                res += 1;
            }
        }
        return res;
    }
};

/**
 * Your AuthenticationManager object will be instantiated and called as such:
 * AuthenticationManager* obj = new AuthenticationManager(timeToLive);
 * obj->generate(tokenId,currentTime);
 * obj->renew(tokenId,currentTime);
 * int param_3 = obj->countUnexpiredTokens(currentTime);
 */

/*
https://leetcode.cn/submissions/detail/400750982/

执行用时：
120 ms
, 在所有 C++ 提交中击败了
23.83%
的用户
内存消耗：
29.5 MB
, 在所有 C++ 提交中击败了
40.65%
的用户
通过测试用例：
90 / 90
*/
